src/widgetastic/browser.py

```diff
@@ -283,31 +283,38 @@ def wait_for_element(
                      also checks visibility.
             timeout: How long to wait for.
             delay: How often to check.
-            exception: If True (default), in case of element not being found an exception will be
-                       raised. If False, it returns False.
+            handle_exception: If True (default), NoSuchElementExceptions will be ignored until the wait is complete
+                              If false, the first NoSuchElementException will be immediately raised
             ensure_page_safe: Whether to call the ``ensure_page_safe`` hook on repeat.
 
         Returns:
             :py:class:`selenium.webdriver.remote.webelement.WebElement` if element found according
-            to params. ``None`` if not found and ``exception=False``.
+            to params.
 
         Raises:
             :py:class:`selenium.common.exceptions.NoSuchElementException` if element not found and
-            ``exception=True``.
+            ``handle_exception=False``.
+            :py:class:`wait_for.TimedOutError` if element not found and ``handle_exception=True``
         """
-        try:
-            result = wait_for(
-                lambda: self.elements(locator, parent=parent, check_visibility=visible,
-                                      check_safe=ensure_page_safe),
-                num_sec=timeout, delay=delay, fail_condition=lambda elements: not bool(elements),
-                fail_func=self.plugin.ensure_page_safe if ensure_page_safe else None)
-        except TimedOutError:
-            if exception:
-                raise NoSuchElementException('Could not wait for element {!r}'.format(locator))
-            else:
-                return None
-        else:
-            return result.out[0]
+        def _element_lookup():
+            try:
+                return self.elements(locator,
+                              parent=parent,
+                              check_visibility=visible,
+                              check_safe=ensure_page_safe)
+            except NoSuchElementException:
+                if not handle_exception:
+                    raise
```

Comment:
I agree to a point, but wanted to retain some of the original method's behavior with the non-default path. If the caller explicitly passes `handle_exception=False`, they must want the old function's behavior, in which case it raises immediately.  The call to wait_for below intentionally does not include handle_exception, so that exceptions other than NoSuchElementException that occur during execution will not be hidden by wait_for and will be allowed to bubble all the way up to the caller.

I might switch the logic here to remove the use of `not`.

Reply:
I see, but previous behavior was basically broken, method didn't do what it was supposed to. And your approach will retain backwards compatibility with that buggy behavior. As i pointed in 
https://github.com/RedHatQE/widgetastic.core/pull/131#issuecomment-446634134 , behavior is changed anyways. Do we really need such compatibility? Is there any real use case, when user wanted to use `wait_for_element` not to wait for element but fail on first attempt if it's not found? Why not to use `browser.element()` directly for such case?

Reply:
Oh, and since method's signature is changed - user should update his code anyway if he wants old behavior. But if user wants to get element without waiting for it (and receive NoSuchElementException otherwise) - he should use `browser.element()` instead. So i'd vote for not combining 2 different use cases in one helper.